src/ttlv/fast_scan.rs

```diff
@@ -0,0 +1,1009 @@
+//! Scanning TTLV elements really fast.
+
+use core::fmt;
+
+use super::types::Tag;
+
+//----------- FastScanner ----------------------------------------------------
+
+/// A fast TTLV scanner.
+///
+/// In most cases, errors are highly unlikely.  This scanner optimizes for
+/// this by eliding error handling and pinpointing facilities -- it detects
+/// errors without tracking any details of them.  In some cases, it will
+/// delay errors, returning garbage data instead of [`Err`].  If scanning
+/// with [`FastScanner`] fails, a different scanner needs to be used to
+/// ascertain the position and cause of the error.
+#[derive(Clone)]
+#[must_use = "Call 'FastScanner::finish()' to finish scanning"]
+pub struct FastScanner<'a> {
+    /// The currently remaining input.
+    remaining: &'a [[u8; 8]],
+
+    /// A failure word.
+    ///
+    /// This is non-zero if an error was detected and delayed.
+    failure_word: u64,
+}
+
+impl<'a> FastScanner<'a> {
+    /// Construct a new [`FastScanner`].
+    pub fn new(input: &'a [u8]) -> Result<Self, FastScanError> {
+        let num_blocks = input.len() / 8;
+        if !input.len().is_multiple_of(8) {
+            // The input contains a partial block.
+            return Err(FastScanError::assert());
+        }
+
+        let ptr = input.as_ptr().cast::<[u8; 8]>();
+        let input = unsafe { core::slice::from_raw_parts(ptr, num_blocks) };
+        Ok(Self::from_blocks(input))
+    }
+
+    /// Construct a new [`FastScanner`] from a block slice.
+    pub const fn from_blocks(input: &'a [[u8; 8]]) -> Self {
+        Self {
+            remaining: input,
+            failure_word: 0,
+        }
+    }
+
+    /// The remaining input to the scanner.
+    pub const fn remaining(&self) -> &'a [[u8; 8]] {
+        self.remaining
+    }
+
+    /// Whether the scanner is finished.
+    pub const fn is_empty(&self) -> bool {
+        self.remaining.is_empty()
+    }
+
+    /// Whether an error has been delayed.
+    ///
+    /// This should only be checked if the consistency of the previously
+    /// scanned data is important, e.g. if acting on it.  It can be ignored
+    /// while parsing.
```

Comment:
```suggestion
    /// Call this function if you want to use scanned data prior to calling
    /// [`finish()`].
    ///
    /// To optimize the scanner, some kinds of error will not cause the
    /// scanner to fail until [`finish()`] is called, and so until that point it
    /// is possible that the data returned by prior scanning may be
    /// incorrect.
    ///
    /// This function checks for an error now, rather than waiting until
    /// [`finish()`] is invoked.
```